### ISSUE TITLE
Inconsistent handling of shortened versions

### ISSUE THREAD

**Issue by adamjstewart** [spack/spack#8957]
Spack seems to be inconsistently handling shorted versions.

### `depends_on`

:heavy_check_mark: **Edited by @haampie 2021-09-22**: the issue here is no longer relevant, because `depends_on('python@2.7')` means the python version is in the closed-open range [2.7, 2.8).

When used in a `depends_on` statement, if I say:
```python
depends_on('python@2.7')
```
Spack will install _exactly_ Python 2.7, which will crash because 2.7 doesn't exist. I would instead expect it to install the latest 2.7.X. This is a long-standing bug documented in #921 and #1035.

As of now, the workaround has been to explicitly turn this into a version range:
```python
depends_on('python@2.7.0:2.7.999')
```
This works, but is unnecessarily wordy.

### `when=`

:heavy_check_mark: **Edited by @haampie 2021-09-22**: This issue is fixed, since we can write `when='@0.2:0.2.0'`, and the meaning is consistent with `depends_on`.

When used in a `when=` argument, if I say:
```python
depends_on('r@3.4.0:3.4.9', when='@0.2')
```
this will apply to both versions 0.2 and 0.2.2, as documented in #8950. Similarly, if I say:
```python
patch('foo.patch', when='@3.2')
```
this patch will be applied to both versions 3.2 and 3.2.1, as documented in #8320 and #8432.

As of now, there is no workaround for this problem. This may be fixed by #8320.

### Compiler versions

This one is the worst of all. Let's say I have a GCC 7.2.0 compiler in my `compilers.yaml`. There are many other GCC compilers, but this is the only GCC 7 compiler I have. If I say:
```console
$ spack install zlib %gcc@7
```
Spack will correctly deduce that I want it to use my GCC 7.2.0 compiler, but it will resolve the compiler as `%gcc@7` in my spec, and install to the `darwin-highsierra-x86_64/gcc-7` directory. The installation succeeds, but if I rerun:
```console
$ spack install foo %gcc@7.2.0 ^zlib
```
Spack needs to reinstall `zlib` with the new GCC 7.2.0 compiler. I believe this bug is undocumented. In fact, I think it used to work properly. I used it all the time when I had multiple GCC compilers installed and I wanted to specify one.

The workaround is to use a version range, like `%gcc@7:`.

### Proposal

All of these situations should be handled identically. Otherwise, it's confusing what behavior to expect. I propose the following behavior.

If a `package.py`/`compilers.yaml` contains both 3.2 and 3.2.1, 3.2 should refer to _exactly_ 3.2 only.

If a `package.py`/`compilers.yaml` only contains 3.2.1, 3.2 should refer to any 3.2.X version.

@tgamblin @baberlevi @PDoakORNL @cyrush @davydden @goxberry @eschnett

**Comment by davydden** [comment]
> All of these situations should be handled identically.

fully agree

> If a package.py/compilers.yaml contains both 3.2 and 3.2.1, 3.2 should refer to exactly 3.2 only.
> If a package.py/compilers.yaml only contains 3.2.1, 3.2 should refer to any 3.2.X version.

personally I don't like the fact that the meaning depends on the case at hand (i.e. presence of absence of `3.2`). 

I would prefer that everything is done using one set of rules. Namely that `3.2` means versions `3.2` and if it does not exist, there is an error. Same with compilers, `%gcc7` should not work. Yes, if you need to cover a range in `when` or `depends_on`, it would get a bit lengthy `2.0:2.99`.

IMO all version comparisons/constraints/etc should have once set of rules using `Version` class and its friends. A single set of rules on the topic of _versions_ would get the least number of surprises for users and package maintainers/developers.

**Comment by adamjstewart** [comment]
> A single set of rules on the topic of versions would get the least number of surprises for users and package maintainers/developers and will be uniform across the Spack.

And yet I still see PRs every month that try to do:
```python
depends_on('python@2.7')
```
or:
```python
depends_on('python@3')
```
It's pretty well documented that this syntax doesn't work in Spack, but it's the most natural syntax for the majority of users. All of the package installation docs say "requires Python 2.7", so why shouldn't Spack support that syntax?

The changes I'm proposing won't change how hashes work. If you install `foo@1.2` and it defaults to `foo@1.2.3`, then you later add `foo@1.2` to the package, nothing will break. Spack will already know that you installed `foo@1.2.3`, not `foo@1.2`. What I'm proposing is a change in the way concretization determines which version to install. 

The vast majority of packages don't have this problem. All versions have the same number of components. For these packages, if I ask to install `python@2.7`, do you really think the best behavior to is to try download _exactly_ 2.7 even though it isn't in the package? I think Spack can be smarter than that.

**Comment by davydden** [comment]
> if I ask to install python@2.7, do you really think the best behavior to is to try download exactly 2.7 even though it isn't in the package?

no, of course not, what I propose is to throw an error.

**Comment by healther** [comment]
There are two levels we should think about here.

One is the user facing level, at this point it really isn't up to us to tell users how they have to think! We consistently see that `@1` is expected to match everything from `@1.0.0.beta` up to (but not including) `2.0`. Even though I agree with @davydden that it is problematic that resolution of `@1` depends on whether a `version('1'` does or does not exist, it is too natural to expect it to work and therefore `spack` should support this. 
We may be able to train our users to do it our way ™️ (i.e. provide an error and suggest the correct version), but is this really worth it? Or even is it less work than just implementing the matching correctly?

But on the other hand we have how we internally handle versions. Here we are completely free to do as we want (with sensible constructions being easier to maintain, but there is no "user"-catchall argument). Here I would say that we should _always_ use the versions as they are declared in `package.py`. I.e. all spack-generated folder structure should contain the version from `package.py` and NOT the user provided strings.

@adamjstewart how is "being able to install versions even if they don't exist in the `package.py`." supposed to work? 
The one situation that I can think about right now is, that we install a package from one branch of spack and then switch to a branch that does not have the installed version of this package. But in that case the database "knows" the version, so no need to go with the user-provided string for internal use

**Comment by adamjstewart** [comment]
> @adamjstewart how is "being able to install versions even if they don't exist in the package.py." supposed to work?
```console
$ spack install python@3.6.6
...
==> Installing python
==> Warning: There is no checksum on file to fetch python@3.6.6 safely.
==>   Fetch anyway? [y/N] y
==> Fetching with no checksum.
  Add a checksum or use --no-checksum to skip this check.
==> Fetching http://www.python.org/ftp/python/3.6.6/Python-3.6.6.tgz
######################################################################## 100.0%
...
```

**Comment by healther** [comment]
Ah okay, I see. Not sure, how useful this use case really is... Changing `package.py` is a rather low bar to add new versions isn't it?

**Comment by adamjstewart** [comment]
@healther it becomes more annoying when you're "installing" an external version that may be too old to download/checksum.

**Comment by PDoakORNL** [comment]
The impression I formed when working on #8320 and reading Version is the following:

Discounting letters in versions in an attempt to be able to deal with as many projects version systems as possible.
Versions are a number system a.b.c where a,b,c are separate positive integers. a+1.b.c > a.x.y for any b,c,x,y. a.b+1.c > a.b.y for any c,y. This relationship continues out any number of points a,b,c,d...

In my opinion `depend` and `install` imply satisfaction semantics where `a.b` is expanded as
`a.b = [a.b.0 : a.b+1.0)` to do comparison with the local set of versions. `[` and `(` having their mathematical interval meanings of inclusive and exclusive respectively. The local versions are expanded with .0's until all local versions have equivalent depth so a concrete a.b is checked as a.b.0.

if `a.b.x` are present locally the greatest `a.b.g` satisfies the depends or the install. An install should to stop and inform me I already have that greatest `a.b.g` and package depends_on command would be satisfied and the spec of the package should be designated to `a.b.g` 
If not a.b.x's are present locally the greatest known version in the package in the expanded a.b should be installed which could be a.b.

If no such version in the expanded `a.b` interval is installed locally and `--unsafe-remote-versions` is set spack should look for the greatest remote version.  

the when= clause implies a equality semantics which are accomplished with the exception that the left side be expanded by with .0 if the right side has more points.  So for a.b == x.y.z  must have a==x, b==y, 0==z. 

So if you want to install mpich 3.2 when 3.2.1 is know you must type
```
spack install mpich@3.2.0
```
because this expands to [3.2.0, 3.2,1) and when checking for satisfaction the version 3.2 is expanded to 3.2.0 which satisfies mpich@3.2.0

I think this removes the need for the different behavior based on the contents of package.py. The exceptional notation need only appear when the exceptional step insisting on an unpatched version is desired.
The `patch` command works in a sane manner.

At least if I I have this correct.
And fixes the contains vs. satisifies issue of #8320.

**Comment by sethrj** [comment]
From the user perspective (and I guess I've constructed a few small `packages.py` myself) I'd be strongly in favor of intelligently resolving unspecified versions. If I want to install the latest version of python 2, I can't run `spack install python@2` -- or at least, I can but it will print a warning about a missing checksum and then install `2.0.0`. So I have to use `spack info` myself and figure out which is the latest available version of python 2.x.

Here's a thought on a possible solution. Currently, spack resolves `spack spec python` using the `preferred=True` argument. What if `preferred=` could take a version specification such as `@2`? So if there's no specific version match, the closest `preferred` version number would be returned.

**Comment by adamjstewart** [comment]
@odoublewen found a new case of this in #11371:
```console
$ spack python
Spack version 0.12.1
Python 3.7.0, Darwin x86_64
>>> from spack.version import *
>>> ver('2019').satisfies(ver('2019.1'))
False
>>> ver('2019').satisfies(ver('2019.1:'))
True
>>> ver('2019').satisfies(ver(':2019.1'))
True
```
In this example, version 2019 satisfies both `:2019.1` and `2019.1:`. This means that there is no way to patch 2019.1 and up without also patching 2019. I have not yet found a workaround for this bug.

**Comment by adamjstewart** [comment]
I'm starting to get a better understanding of this most recent bug. `ver('2019')` is a `Version` object and `ver('2019.1:')` is a `VersionRange` object. In order to compare these two types of objects, we coerce `Version('2019')` into `VersionRange('2019:2019')`. Thus, `2019.1:` overlaps with `2019:2019`.

**Comment by mwkrentel** [comment]
@tgamblin  @adamjstewart  @scheibelp 

Would it be possible to prioritize this issue and at least fix some of
the more egregious cases?

I just fixed some patches for intel-tbb (#13932).  TBB has
multiple-length version numbers: 2018, 2018.1, 2019, 2019.1, etc.
There are patches that apply to :2019.4, 2019.5:, etc.

I try `spack install intel-tbb @2019`, but both of these are TRUE:

```
2019 satisfies :2019.2
2019 satisfies 2019.5:
```

WTF!?  How is it possible to use version ranges when versions don't
even obey a linear order?

I guess this problem affects any package that uses multiple-length
version numbers.  So, there are likely packages with broken version
ranges that don't know it.

Btw, if I were the maintainer of TBB, I would want to rename 2019 to
2019.0, at least until this is resolved.

----------

This is one of those cases where spack outsmarts itself.  Sometimes
2019 means exactly 2019 and sometimes it means 2019.0:2019.99.  And
the complicated interpretation breaks the simple case.

I would be ok with changing this to: 2019 means exactly 2019 and if
you want the other meaning, then write 2019.0:2019.99 or 2019.* or
some such.

At least until the complicated meaning doesn't break the simple case.

**Comment by healther** [comment]
Every time I come back to this kind of issue I have the (somewhat crazy idea) of introducing special characters in the versioning scheme. This is not particularly well thought through, so if you see obvious holes in it tell me and we forget about it again.

What about introducing special characters in the version-scheme? What I would propose is that we have by default always a `major.minor.patch` version.
- If the package does not come with that we expand all shorter versions with one special character (let's say we use 'A' for this). So `2.0`, `20191201`, `1.6.79` would become `2.0.A`, `20191201.A.A`, `1.6.79`.
- If we want to depend on a range we can replace `minor.patch` and/or `patch` by a different special character (lets say we use `?` for this). The `depends_on('python@2.7.?')` for python27 dependencies.
This does not fix everything, but should take care of most of the packages we currently have and I _hope_ that it should be easier to implement such an augmentation scheme.

I think we need to have a general solution to this, independent of a new concretiser as the conflict originating from
```
version('2.7', ...)
version('2.7.2', ...)
```
will not go away...

**Comment by scheibelp** [comment]
@mwkrentel I think https://github.com/spack/spack/issues/8957#issuecomment-560183994 is a different issue: 

* The original post here asks for `python@2.7` to be interpreted as `2.7:2.7.999` (approximately)
* You are asking that a patch when clause like `@2015` be interpreted as `@2015:2015` (I think - I'm inferring that from your post)

https://github.com/spack/spack/issues/13831#issuecomment-558854712 is an example of how to resolve it and https://github.com/spack/spack/pull/13906#issuecomment-558919745 gets into the details of why `@2015` is ambiguous (and in some sense why your request and the OP are at odds, not to say it isn't worth trying to get the best of both worlds).

It would be perfectly valid to file a separate issue on this. If you agree I will collapse https://github.com/spack/spack/issues/8957#issuecomment-560183994 and this comment.

**Comment by adamjstewart** [comment]
@scheibelp I think https://github.com/spack/spack/issues/8957#issuecomment-560183994 fits within the scope of this issue. I myself reported the same thing in https://github.com/spack/spack/issues/8957#issuecomment-493704796

This issue pertains to the numerous ways in which Spack handles shortened versions, usually inconsistently. This includes `patch`, `depends_on`, `when`, etc.

**Comment by scheibelp** [comment]
Thanks for pointing that out, I have edited https://github.com/spack/spack/issues/8957#issuecomment-560539408 to remove the suggestion to split this off. 

That being said, regarding

> How is it possible to use version ranges when versions don't even obey a linear order?

https://github.com/spack/spack/issues/8957#issuecomment-560539408 (the part that is not crossed out) contains pointers to explanations of how to write `when` constraints that apply the patches for the appropriate specs. So at this point my understanding that this a consistency problem in the sense that it is confusing for users (rather than than a limitation on what it's possible to express in a `package.py` file).

**Comment by scheibelp** [comment]
While the handling is inconsistent across *all* instances of specs (command line, `when` constraint, etc.) it is consistent with the following set of rules:

* if the spec is in a `when` constraint, it is interpreted broadly (for reasons mentioned in https://github.com/spack/spack/pull/13906#issuecomment-558919745)
* if the spec is anywhere else (command line, `depends_on`), then it is interpreted strictly. For command-line specs that is because of https://github.com/spack/spack/issues/8957#issuecomment-412296498

This could be relaxed for

* `depends_on`
* compiler versions

without violating https://github.com/spack/spack/issues/8957#issuecomment-412296498 but getting that right would take some effort

**Comment by mwkrentel** [comment]
@tgamblin  @adamjstewart  @scheibelp  @becker33 

Well, I **thought** about opening a new issue and pretending not to
remember this one, just to see if that would get more attention.  :-)
But clearly, Adam reported the same issue above.

And to be clear, this is not an issue of interpreting `2019` inside
`when()`.

I tried `spack install intel-tbb @2019` and the concretizer picks rev
exactly 2019, not 2019.8.  (This much is correct.)  But as Adam
reported, the problem is that `2019` satisfies both of these:

```
patch(..., when='@2019.6:`)
patch(..., when='@:2019.2')
```

So, rev 2019 is both >= 2019.6 and <= 2019.2.

Basically, this makes it impossible to use version ranges for any
package that has variable-length versions.  There is no valid order
where this makes sense.

IMO, this is a clear bug, seperate from how you want to interpret
`when='@2019'`.  (One reason I always specify a full version inside
when.)

I was hoping this would be a small and isolated fix and could be done
before working out the full details of how to interpret shortened
version names.

Btw, does the full fix for this fall under the domain of the new
concretizer?  Is that what we're waiting on?

**Comment by scheibelp** [comment]
I was going to suggest creating a constraint like `2019.2:2019.99` (which I thought was analogous to the suggestion in https://github.com/spack/spack/issues/13831#issuecomment-558854712) but after trying that myself I see it doesn't work.

So the problem you are discussing is more than an inconsistency: it's a gap in the model's ability to properly specify constraints (so I think in fact https://github.com/spack/spack/issues/8957#issuecomment-560568209could be considered separate). I don't think addressing your issue specifically will be very difficult although it likely won't address the broader issue here.

**Comment by mwkrentel** [comment]
Ok, I spun off this last problem in #13961.  It's small, isolated and
hopefully can be fixed without needing the full solution for how to
treat shortened version numbers.

**Comment by mwkrentel** [comment]
The separate issue (bug) that I spun off in #13961 is now fixed,
although some constraints still need an explicit `strict=True`.
Thanks to @scheibelp !

----------

But I'd like to call out one difficult case here for the time when people
think about a new design for this issue.

Consider a package with variable length version numbers: `2.1`,
`2.1.0` and `2.1.1`.  Currently, it's not possible to write a `when`
clause that applies to exactly `2.1` and not include `2.1.0`.

It's the `0` that's troublesome here.  If you only had `2.1` and
`2.1.1`, then `when='@2.1:2.1.0'` would work.

So, part of the cost of having `when='@2.1'` meaning `2.1.0:2.1.999`
is that you have to invent a new syntax to say, "please interpret
`2.1` as exactly `2.1`, not `2.1.*`."

Maybe something like `@=2.1`.

**Comment by haampie** [comment]
Copying my issue from elsewhere since it was a duplicate:

Given:
1. Two system compilers with specs `gcc@10.2.0-myidentifier` and `gcc@10.3.0-myidentifier`
2. A package `x` depending on `y`, where `x` conflicts with `gcc@10.3.0`, and also conflicts with `y@2.0.0`
3. And a system version of `y` with the spec `y@2.0.0-myidentifier`

```spack spec x``` concretizes to

```
x@1.0.0%gcc@10.3.0-myflag arch=linux-ubuntu20.04-zen2
    ^y@2.0.0-flag%gcc@10.3.0-myflag arch=linux-ubuntu20.04-zen2
```

not satisfying either of the constraints, which is a bug, because the concrete output spec satisfies the specs of the conflicts:

```python
>>> x = spack.spec.Spec('x')
>>> x.concretize()
>>> x
x@1.0.0%gcc@10.3.0-myflag arch=linux-ubuntu20.04-zen2 ^y@2.0.0-flag%gcc@10.3.0-myflag arch=linux-ubuntu20.04-zen2
>>> x.satisfies(spack.spec.Spec('%gcc@10.3.0'))
True
>>> x.satisfies(spack.spec.Spec('^y@2.0.0'))
True
```

This is because Spack special-cases "concrete versions", where it arbitrarily switches interpretation of the meaning of `x.y.z` as the closed-open range `[x.y.z, x.y.z+1)` to the exact version `x.y.z`.

IMHO Spack should not do this, and uniformly interpret `x.y.z` as a range.

If I understand correctly this behavior was put in place just to support packages with weird numbering schemes `1.0`, `1.0.1`, `...`, `1.0.4`, `1.1`, `...`.

But this is not a good justification, cause in those packages we can simply work around the issue with ranges too, such as:

```python
depends_on('x@1.0:1.0.0')  # only allow 1.0, not 1.0.1, ...
conflicts('x@:1.0.0')  # support 1.0.1, 1.0.2, ... but not 1.0
```

---

To reproduce, create a repo `spack repo create example example`, put these two packages in there, and use the spack.yaml file in the same dir containing the repo:

```python
class X(Package):
    homepage = "localhost"
    has_code = False

    version('1.0.0')

    depends_on('y')

    conflicts('%gcc@10.3.0')
    conflicts('^y@2.0.0')
```

```python
class Y(Package):
    homepage = "localhost"
    has_code = False
    
    version('2.0.0')
    version('1.0.0')
```

```yaml
spack:
  specs:
  - x
  repos:
  - ./example
  packages:
    y:
      externals:
      - spec: y@2.0.0-myidentifier
        prefix: /path/to/somewhere
  'compilers:':
  - compiler:
      environment: {}
      extra_rpaths: []
      flags: {}
      modules: []
      operating_system: ubuntu20.04
      paths:
        cc: cc
        cxx: CC
        f77: ftn
        fc: ftn
      spec: gcc@10.2.0-myidentifier
      target: x86_64
  - compiler:
      environment: {}
      extra_rpaths: []
      flags: {}
      modules: []
      operating_system: ubuntu20.04
      paths:
        cc: cc
        cxx: CC
        f77: ftn
        fc: ftn
      spec: gcc@10.3.0-myidentifier
      target: x86_64
```

**Comment by haampie** [comment]
> So if you want to install mpich 3.2 when 3.2.1 is know you must type `spack install mpich@3.2.0`

No you don't, you can do `spack install mpich@:3.2.0`, which shows more of the intent to exclude `3.2.1` but include `3.2`, it does not read as "install non-existing version 3.2.0" anymore. You can even do `spack install mpich@3.2:3.2.0` to take any release from the range [3.2, 3.2.1).

> Consider a package with variable length version numbers: 2.1, 2.1.0 and 2.1.1. Currently, it's not possible to write a when
clause that applies to exactly 2.1 and not include 2.1.0.

Do we know of cases where `2.1` and `2.1.0` are separate releases? Seems like an edge case not worth focusing on. But I do agree, this would be the only time when we actually *require* new spec syntax because 2.1 and 2.1.0 map to the same point and can hence not be separated with ranges. If we do get syntax for this, it's useful if you don't like to write `depends_on('mpich@3.2:3.2.0')` but rather write `depends_on('mpich@=3.2')` so that others don't pull their hairs out trying to understand what you mean.

---

However, I'd like to bring forward the use case where flags are used in specs listed in compilers/packages.yaml files. In particular when working with Cray, they may have many different development toolkits on different versions, but the compiler may not have been updated between two releases of toolkits. Whenever you're using a toolkit at a particular version, you want to module load cray-xyz packages from the same toolkit.

The only sensible way to do this is to have multiple entries for say `gcc@[compiler version]-[toolkit version]` e.g. `gcc@10.3.0-21.05` and `gcc@10.3.0-21.09`, and have the `-21.05` spec load a module `cdt/21.05` together with the compiler, and have `-21.09` load `cdt/21.09`.

If there is a package defining a `conflict('%gcc@10.3.0')`, then a compiler with spec `gcc@10.3.0-21.05` does satisfy `gcc@10.3.0` as a compiler spec, but the concretizer chooses to believe `gcc@10.3.0` is an exact version and hence it refuses to respect the conflict. The only workaround is to write `conflicts('%gcc@10.3.0:10.3.0.0')` which is very awkward... It doesn't make sense for GCC like it *does* make sense to mpich with its odd 3.2, 3.2.1 versioning.

Reading through this thread, it seems we've already moved away from `depends_on(python@2.7)` being the literal version 2.7, and instead decided it means [2.7, 2.8). Shouldn't we inform the concretizer about this too?

**Comment by haampie** [comment]
To make the clingo concretizer interpret `x.y` as `[x.y, x.y+1)` consistently, this seems the minimal change, but it fails tests:

```diff
diff --git a/lib/spack/spack/solver/asp.py b/lib/spack/spack/solver/asp.py
index 790e545878..37788c41ae 100644
--- a/lib/spack/spack/solver/asp.py
+++ b/lib/spack/spack/solver/asp.py
@@ -982,11 +982,7 @@ class Body(object):
         if spec.compiler:
             clauses.append(f.node_compiler(spec.name, spec.compiler.name))
 
-            if spec.compiler.concrete:
-                clauses.append(f.node_compiler_version(
-                    spec.name, spec.compiler.name, spec.compiler.version))
-
-            elif spec.compiler.versions:
+            if spec.compiler.versions:
                 clauses.append(
                     fn.node_compiler_version_satisfies(
                         spec.name, spec.compiler.name, spec.compiler.versions))
@@ -1230,12 +1226,6 @@ def define_version_constraints(self):
                 if v.satisfies(versions)
             ]
 
-            # This is needed to account for a variable number of
-            # numbers e.g. if both 1.0 and 1.0.2 are possible versions
-            exact_match = [v for v in allowed_versions if v == versions]
-            if exact_match:
-                allowed_versions = exact_match
-
             # generate facts for each package constraint and the version
             # that satisfies it
             for v in allowed_versions:
```

**Comment by adamjstewart** [comment]
Remaining version inconsistencies (if they even still exist) are restricted to compilers, which will soon become formal dependencies. I think we can close this issue. Thanks @haampie for so many great fixes to version handling!
